File: Top_Interview_150/Array_String/cpp/14.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    string longestCommonPrefix(vector<string> &strs)
    {
        sort(strs.begin(), strs.end());
        string left = strs[0];
        string right = strs[strs.size() - 1];

        int length = min(left.size(), right.size());
        string result = "";
        for (int i = 0; i < length; ++i)
        {
            if (left[i] == right[i])
                result += left[i];
            else
                return result;
        }

        return result;
    }
};

File: Top_Interview_150/Array_String/python/14.py
from collections import defaultdict
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        print(sorted(strs))
        d = defaultdict(int)
        n = len(strs)

        for word in strs:
            pref = ""
            for c in word:
                pref += c
                d[pref] += 1

        ans = ""
        for word, count in d.items():
            if count == n:
                ans = max(ans, word, key=len)

        return ans


solution = Solution()
print(solution.longestCommonPrefix(strs = ["flower","flow","flight"]))
